Optimize CScript.FindAndDelete

gavinandresen opened the pull request:
The FindAndDelete function could move the same bytes multiple times if the signature being deleted appeared in the script multiple times.

These two commits add a unit test for FindAndDelete and optimize it so that bytes are moved at most once.

gavinandresen committed:
Optimize CScript::FindAndDelete

CScript::FindAndDelete, called by the CHECKSIG Script opcodes, could
end up moving the bytes of the Script multiple times if there were
multiple matches for the Find.

This commit tweaks the algorithm so it never moves bytes in the Script
more than once per call to FindAndDelete.

gavinandresen committed:
Unit test for CScript::FindAndDelete

kazcw commented on src/script/script.h:
```diff
@@ -571,16 +571,34 @@ class CScript : public CScriptBase
             {
-                pc = erase(pc, pc + b.size());
+                // Lazy copy-in-place if there is a match:
+                while (copied != pc-nShift) {
```
A test for !nShift here would prevent copying everything upto the first match to itself

kazcw commented on src/script/script.h:
```diff
@@ -571,16 +571,34 @@ class CScript : public CScriptBase
+
+        if (nShift > 0)
+        {
+            while (copied != pc-nShift) {
```
I think this changes the result of FindAndDelete for invalid CScripts that end in a pushdata without enough data. Replacing "pc" with "end()" would keep the original behavior in this case.

sipa commented:
Concept ACK on the improvement, but there are maybe a few more edge cases to test:
* Removing a byte sequence that spans multiple opcodes
* Trying to remove a sequence that only occurs in the middle of opcodes (which doesn't work)
* Trying to remove "OP_0 OP_1" from "OP_0 OP_0 OP_1 OP_1" leaves "OP_0 OP_1" (because we don't go backward after a replacement)

gavinandresen commented:
Added unit tests as suggested by @sipa.
Fixed to be compatible with invalid CScripts that end with partial PUSHDATA's (and added unit tests) as suggested by @kazcw 

And saved a couple of lines of code (and probably made faster, but copying is never done in normal operation with reasonable transactions) by switching to using std::copy instead of an explicit loop.

Thanks for the code reviews!

TheBlueMatt commented on src/test/script_tests.cpp:
```diff
@@ -1051,4 +1051,121 @@ BOOST_AUTO_TEST_CASE(script_GetScriptAsm)
+    // This is an odd edge case: strip of the push-three-bytes
+    // prefix, leaving 02ff03 which is push-two-bytes:
+    s = ScriptFromHex("0302ff030302ff03");
+    d = ScriptFromHex("03");
```
I'd prefer to not have tests that check that FindAndDelete do exactly what we currently do when called against invalid scripts when we dont require that in any callers.

gavinandresen replied:
RE: behavior with invalid Scripts: I don't care one way or the other; seems safer to behave exactly the same, since this is consensus-critical code.

dcousens commented on src/script/script.h:
```diff
@@ -571,16 +572,30 @@ class CScript : public CScriptBase
             {
-                pc = erase(pc, pc + b.size());
+                // Lazy copy-in-place if there is a match:
+                if (nShift == 0) copied = pc;
```
Isn't `nShift` always `== 0` at the beginning?   Why bother initializing `copied` above?

gavinandresen replied:
Why initialize copied instead of just eliminating the if/else and always copying? Because std::copy behavior is undefined if the first and last arguments point to the same element.

kazcw replied:
Why would that be? The Standard's definition of std::copy doesn't seem to say anything that would prohibit an empty range.

gavinandresen replied:
The standard says:

"result shall not be in the range [first,last)."

kazcw replied:
`last` is not in the range `[first,last)`

gavinandresen replied:
@kazcw :  The first time a match is found, copied will be begin(). nShift will be zero, and pc will be pointing into the array somewhere.

If we removed copied = pc and just did:
   ```copied = std::copy(copied+nShift, pc, copied);``

... then first is copied, last is pc, and result is copied.
copied IS in the range [copied, pc)

sipa commented:
Superseded by #7907.
